### DISCUSSION TITLE
python environment

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#28] [General]
Let's start a discussion here. I'm sure there will be more on this.

Sorry for any confusion on why a python env is needed.
First of all, python env is a developer tool and won't be needed for end-user that just wants to get this repo and put it on his board. So it is up to you if you need it while coding. I do some coding on my desktop when no hardware interaction is needed so I must keep the python isolated from other projects. I also havve a second dev RPi that's used whhen I need to script things that might broke the OS... So this why I can quicky rebuild the python as I don;t need to remember what I installed...

If you do code straight on the board, the only reason to use python env is to get the requitements for someone to use the software. To do this:
- go inside env and use pip3 to install needed packages abd nake sure everythinhg works
- generate the requirements.txt file containing dependencies
```pip3 freeze > requirements.txt ```

You said you added some packages for lichess, you might want to update requirements,txt file

Now anyone who wants to deploy the Centaur mod will just clone the repo and use pip to install all the dependencies.

```
git clone <repo>
cd DGTCentaur
pip3 install -r requirements.txt
```
 Our code should work on the board witrhout an env. Basically the board is its environment.

Paths. I think this is something Luca is looking into it. We need to have a config where a base directory (install directory) is mentioned and then inside the code we just use "<install_dir>/dgt_centaur_mods/...." as a path. Till then everything is manual :) and everyone adapts.

I use this hack sometime to see the working directory from python perspective:
```
import os

print(os.getcwd())
```
Put this in a getcwd.py and place it like a probe where you need to see how python would see the wd. Call it python3.7 getpwd.py

I'd say Centaur software to remain in /home/pi and we should never put anything inside. Then DGTCentaur will be our playground and anyone can decide where they want to put it (/opt, /home/pi, etc.). Then we should have an install script that will taka care on setting up the RPi to work (enable serial, enable spi, install python dependencies on the system, etc). Of coursre pernissions considered,,,
Luca is this something you're looking into? I know you were mentioning something about this.

Hope everything makes sense. :)))

**Comment by wormstein** [comment]
To simply answer your quetions in the pull request :)

No, you dfon't have to change the way you call menu.py. We need to work on the paths in the code to make it work anyway you call it.
Like I said up here, in our case the board is the env as no other python software should be running in there.

PYTHONPATH: I'll look into it and see if I can set it up automatically inside th deply script. The script will have to look where you have cloned the repo and then set this up accordingly. I see this usefull for end-user in the install script...

**Comment by EdNekebno** [comment]
Thanks for explaining to this python newb!
It makes a lot of sense - I added the additional dependencies to requirements.txt . 
Most of my confusion is that I hard coded folder paths into things originally, which is obviously a bad idea but works if one person is working on stuff. Now we all have it in slightly different places. 
The configuration file I am definitely onboard with. But my thought is that now we have luca's structure then things like the font in boardfunctions.py can be coded relative (I pushed a change to that one to show what I mean, it should work wherever our code is so long as it is the correct structure)

`font18 = ImageFont.truetype(str(pathlib.Path(__file__).parent.resolve()) + "/../resources/Font.ttc", 18)`

I think it's pretty much all down to me being hacky when I started and now confusing myself :D
